### ISSUE TITLE
Support scheme in service discovery address

### ISSUE THREAD

**Issue by jlcarpenter** [ThreeMammals/Ocelot#617]
I'm using version 11.0.2 of Ocelot and Ocelot.Provider.Consul.  Below is my ocelot.json file.

`   

{
	  "ReRoutes": [
		{
		  "DownstreamPathTemplate": "/v1/vaults",

		  "ServiceName": "testservice",  
		  //"LoadBalancer": "LeastConnection",
		  //"LoadBalancerOptions": {
		  //  "Type": "LeastConnection"
		  //},
		  "UpstreamPathTemplate": "/api/values",
		  "UpstreamHttpMethod": [ "Get" ]
		}
	  ],
	  "GlobalConfiguration": {
		"RequestIdKey":  "OcelotRequestId", 
		"ServiceDiscoveryProvider": {
		  "Host": "localhost",
		  "Port": 8500,
		  "Type": "Consul"
		}
	  }
	}`

When I submit the request, I get the following log information.
`
 	requestId: 8000001a-0004-ff00-b63f-84710c7967bb, previousRequestId: no previous request id, message: Unable to use service Address: http://localhost and Port: 56736 as it is invalid. Address must contain host only e.g. localhost and port must be greater than 0

	dbug: Ocelot.LoadBalancer.Middleware.LoadBalancingMiddleware[0]
		  requestId: 8000001a-0004-ff00-b63f-84710c7967bb, previousRequestId: no previous request id, message: there was an error leasing the loadbalancer, setting pipeline error

	warndbug: Ocelot.LoadBalancer.Middleware.LoadBalancingMiddleware[0]
		  requestId: 8000001a-0004-ff00-b63f-84710c7967bb, previousRequestId: no previous request id, message: There were no services in NoLoadBalancer

	warn: Ocelot.Responder.Middleware.ResponderMiddleware[0]
		  requestId: 8000001a-0004-ff00-b63f-84710c7967bb, previousRequestId: no previous request id, message: Error Code: ServicesAreEmptyError Message: There were no services in NoLoadBalancer errors found in ResponderMiddleware. Setting error response for request path:/api/values, request method: GET

	dbugdbug: Ocelot.Errors.Middleware.ExceptionHandlerMiddleware[0]
		  requestId: 8000001a-0004-ff00-b63f-84710c7967bb, previousRequestId: no previous request id, message: ocelot pipeline finished`

Do I have something misconfigured?

**Comment by TomPallister** [comment]
@jlcarpenter it looks like your service registered in Consul has http:// in the consul address field. Ocelot doesn't support this at the moment. The services just need the host. More info [here](https://ocelot.readthedocs.io/en/latest/features/servicediscovery.html#consul). 

Let me know if that helps.

**Comment by jlcarpenter** [comment]
I think I have removed the http:// from the consul address field.  Now when I call the api I get the following log.  

`
dbug: Ocelot.Requester.Middleware.HttpRequesterMiddleware[0]
      requestId: 80000013-0006-ff00-b63f-84710c7967bb, previousRequestId: no previous request id, message: IHttpRequester returned an error, setting pipeline error
warndbug: Ocelot.Requester.Middleware.HttpRequesterMiddleware[0]
      requestId: 80000013-0006-ff00-b63f-84710c7967bb, previousRequestId: no previous request id, message: Error making http request, exception: System.ArgumentException: Only 'http' and 'https' schemes are allowed.
      Parameter name: value
         at System.Net.Http.HttpRequestMessage.set_RequestUri(Uri value)
         at Ocelot.Request.Middleware.DownstreamRequest.ToHttpRequestMessage()
         at Ocelot.Requester.HttpClientHttpRequester.GetResponse(DownstreamContext context)
: Ocelot.Requester.Middleware.HttpRequesterMiddleware[0]
      requestId: 80000013-0006-ff00-b63f-84710c7967bb, previousRequestId: no previous request id, message: IHttpRequester returned an error, setting pipeline error
warn: Ocelot.Responder.Middleware.ResponderMiddleware[0]
      requestId: 80000013-0006-ff00-b63f-84710c7967bb, previousRequestId: no previous request id, message: Error Code: UnableToCompleteRequestError Message: Error making http request, exception: System.ArgumentException: Only 'http' and 'https' schemes are allowed.
      Parameter name: value
         at System.Net.Http.HttpRequestMessage.set_RequestUri(Uri value)
         at Ocelot.Request.Middleware.DownstreamRequest.ToHttpRequestMessage()
         at Ocelot.Requester.HttpClientHttpRequester.GetResponse(DownstreamContext context) errors found in ResponderMiddleware. Setting error response for request path:/api/values, request method: GET
dbug: Ocelot.Errors.Middleware.ExceptionHandlerMiddleware[0]
      requestId: 80000013-0006-ff00-b63f-84710c7967bb, previousRequestId: no previous request id, message: ocelot pipeline finished
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
      Request finished in 1755.1627ms 500
`

**Comment by TomPallister** [comment]
@jlcarpenter you need DownstreamScheme in your ReRoute config as per the docs..but annoying but how it works at the moment!

**Comment by jlcarpenter** [comment]
Thanks.  I thought that could be removed when using ServiceDiscovery

**Comment by TomPallister** [comment]
Yeh I will leave this issue open to take a look at it.

Rule could be, if you are using service discovery and havent set downstreamscheme then we expect it in the address. If you have set it then we don't expect it in the address.

**Comment by raime910** [comment]
I get the same issue and I dont have the http/https shemes. The weird thing about this is that after getting ServicesAreEmptyError, I wait for 10-15 seconds, send the request again and it will go through. 

```
[17:48:31 INF] Request finished in 334.7116ms 200 application/json; charset=UTF-8
[17:48:31 INF] Request starting HTTP/1.1 GET http://my.docker.project/.well-known/openid-configuration/jwks  
[17:48:31 INF] Invoking IdentityServer endpoint: IdentityServer4.Endpoints.DiscoveryKeyEndpoint for /.well-known/openid-configuration/jwks
[17:48:31 INF] Request finished in 20.0724ms 200 application/jwk-set+json; charset=UTF-8
Loaded '/usr/share/dotnet/shared/Microsoft.AspNetCore.App/2.2.7/Microsoft.IdentityModel.JsonWebTokens.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
[17:48:31 INF] Successfully validated the token.
[17:48:31 INF] requestId: 0HLR22VIBVHF8:00000001, previousRequestId: no previous request id, message: Client has been authenticated for /archives-api/files
[17:48:31 INF] requestId: 0HLR22VIBVHF8:00000001, previousRequestId: no previous request id, message: route is authenticated scopes must be checked
[17:48:31 INF] requestId: 0HLR22VIBVHF8:00000001, previousRequestId: no previous request id, message: user scopes is authorised calling next authorisation checks
[17:48:31 INF] requestId: 0HLR22VIBVHF8:00000001, previousRequestId: no previous request id, message: route is authorised
[17:48:31 INF] requestId: 0HLR22VIBVHF8:00000001, previousRequestId: no previous request id, message:  has succesfully been authorised for /archives-api/{catchAll}.
[17:48:31 WRN] requestId: 0HLR22VIBVHF8:00000001, previousRequestId: no previous request id, message: services were empty for archives-api
[17:48:31 WRN] requestId: 0HLR22VIBVHF8:00000001, previousRequestId: no previous request id, message: Error Code: ServicesAreEmptyError Message: services were empty for archives-api errors found in ResponderMiddleware. Setting error response for request path:/archives-api/files, request method: POST
[17:48:31 INF] Request finished in 733.0776ms 404 

```

**Comment by raman-m** [comment]
> @secret-agent-B commented [on Nov 5, 2019](#issuecomment-549938550)
ServicesAreEmptyError

This is Service Discovery problem vs Consul was fixed by #1670 
- #1670 

Upgrade to the latest version!

**Comment by raman-m** [comment]
> @jlcarpenter commented [on Sep 12, 2018](#issue-359545157)

I'm not sure this bug is still presented in latest versions.
But if it Is, then feel free to reopen with reporting in details for Ocelot v22+ please.
Upgrade to newest version plz!
